### PR TITLE
Adding User Setting Entity and associated modules

### PR DESCRIPTION
**Pull request by willdoran**
This pull request makes the following changes:
- Add User Setting entity

Test checklist:
- [ ] Run tests

- [ ] I certify that I ran my checklist

Fixes ushahidi/platform#2807

Ping @ushahidi/platform

### CODE REVIEW

File: routes/web.php [hunk 1]
```diff
@@ -477,6 +477,16 @@
         // Public access
         $router->get('/', 'UsersController@index');
         $router->get('/{id:[0-9]+}', 'UsersController@show');
+        $router->get('/settings', 'SettingsController@index');
+
+        // Sub-user routes
+        $router->group(['prefix' => '/{user_id:[0-9]+}'], function () use ($router) {
```
Comment (rowasc) [inline routes/web.php]: @willdoran I think the routes need to be inside the hxl feature flag check ((Features::isEnabled('hxl')))

File: src/App/Repository/User/SettingRepository.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	// UserSettingRepository
+	public function existsInUserSetting($user_id)
+	{
+		return (bool) $this->selectCount(compact('user_id'));
+    }
+    
+	public function create(Entity $entity)
+	{
+		$record = $entity->asArray();
+		$record['created'] = time();
```
Comment (rowasc) [inline src/App/Repository/User/SettingRepository.php]: @willdoran why not use a default timestamp instead?

File: tests/integration/users/settings.feature [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                "config_key":"key",
+                "config_value":"value"
+            }
+            """
+        When I request "/users/9999/settings"
+        Then the response is JSON
+        And the response has a "errors" property
+        Then the guzzle status code should be 404
+
+    Scenario: Creating a new Settings on a Stage with the wrong Form ID
```
Comment (rowasc) [inline tests/integration/users/settings.feature]: is this supposed to be `wrong user id` instead?

File: tests/integration/users/settings.feature [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                "priority":1
+            }
+            """
+        When I request "/users/2/Settingss"
+        Then the response is JSON
+        And the response has a "errors" property
+        Then the guzzle status code should be 400
+
+    Scenario: Updating a Settings
+        Given that I want to update a "Settings"
```
Comment (rowasc) [inline tests/integration/users/settings.feature]: I don't think updating a key should be possible, can you explain this one ? I'm probably just extra slow today

Comment (rowasc) [reply tests/integration/users/settings.feature]: ok yea extra slow it it. Ignore.

File: routes/web.php (hunk 1 above)
Comment (willdoran) [reply routes/web.php]: @rowasc 
I don't think so, I think this could legitimately be considered a separate feature since we're likely to use user_settings elsewhere. Would it not deserve its own flag or no?

But I can hide it for now behind that flag with a comment.

File: src/App/Repository/User/SettingRepository.php (hunk 2 above)
Comment (willdoran) [reply src/App/Repository/User/SettingRepository.php]: Only for consistency with other Repos

Comment (willdoran) [reply src/App/Repository/User/SettingRepository.php]: But can change

**Comment by willdoran** [comment]
@rowasc I didn't add the feature flag because I want to see if I can extract the routes easily and place them in a features route file so that we can easily see what is and isn't feature flagged. i'll try that this morning

**Comment by rowasc** [comment]
@willdoran ah that's ok. LMK if you want me to review again

File: app/Http/Middleware/CheckFeature.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Closure  $next
+     * @param  mixed  ...$features
+     * @return mixed
+     */
+    public function handle($request, Closure $next, ...$features)
+    {
+        foreach ($features as $feature) {
+            if (!Features::isEnabled($feature)) {
+                abort(403, "Forbidden.");
```
Comment (rowasc) [inline app/Http/Middleware/CheckFeature.php]: Robbie asked for a 404 when a feature is not enabled. But I'm not sure if we want to do it for everything or just the hdx feature. 

cc @rjmackay can you check this?

File: src/App/Validator/User/Setting/Update.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+	public function __construct(UserRepository $user_repo)
+	{
+		$this->user_repo = $user_repo;
+	}
+
+	protected function getRules()
+	{
+		return [
+			'user_id' => [
```
Comment (rowasc) [inline src/App/Validator/User/Setting/Update.php]: is it github or is there something wrong with the spaces/tabs here?

File: tests/integration/bootstrap/RestContext.php [hunk 7]
```diff
@@ -279,9 +279,9 @@ public function iRequest($pageUrl)
 
 		// Get response object
 		$this->response = $response;
-		// $data = $this->response->getBody(true);
-		// $data = explode("\n", $data);
-		// \Log::info(print_r($data, true));
+		$data = $this->response->getBody(true);
+		$data = explode("\n", $data);
```
Comment (rowasc) [inline tests/integration/bootstrap/RestContext.php]: I'm not sure I understand why we do this here for the print r?

File: app/Http/Middleware/CheckFeature.php (hunk 5 above)
Comment (willdoran) [reply app/Http/Middleware/CheckFeature.php]: Cool we can make it 404 instead with no issue, though it would be nice if this wasn't English it looks weird having a string attached to it. Surely, it can just be 404 and the browser figures it out

File: tests/integration/bootstrap/RestContext.php (hunk 7 above)
Comment (willdoran) [reply tests/integration/bootstrap/RestContext.php]: oh sorry that should be commented out, I forgot to revert it

File: bootstrap/lumen.php [hunk 8]
```diff
@@ -104,6 +102,7 @@
     'namespace' => 'Ushahidi\App\Http\Controllers',
 ], function ($router) {
     require __DIR__.'/../routes/web.php';
+    require __DIR__.'/../routes/features.php';
```
Comment (rjmackay) [inline bootstrap/lumen.php]: Maybe we should split up routes per entity endpoints and require each one within the `web.php` file?

File: app/Http/Middleware/CheckFeature.php [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+	 *
+	 * @param  \Illuminate\Http\Request  $request
+	 * @param  \Closure  $next
+	 * @param  mixed  ...$features
+	 * @return mixed
+	 */
+	public function handle($request, Closure $next, ...$features)
+	{
+		foreach ($features as $feature) {
+			if (!Features::isEnabled($feature)) {
```
Comment (rjmackay) [inline app/Http/Middleware/CheckFeature.php]: TBH I actually preferred this being a simple if statement within routes. My thinking was that feature toggle shouldn't be complicated, and also that routes simply shouldn't exist if the feature is disabled (vs existing but having middleware). 

That said: I'm not strongly attached to my approach. If you/@rowasc prefer this way, then lets do it. As long as we pick a method and stick with it.

File: routes/features.php [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+	// HXL
+	$router->group([
+		'prefix' => 'hxl',
+		'middleware' => ['feature:hxl']
+	], function () use ($router) {
+		$router->get('/', "HXLController@index");
+	});
+
+	// User Settings
```
Comment (rjmackay) [inline routes/features.php]: I kind prefer keeping these within the rest of the routes, so we don't have to guess which file something is in depending on if it has a feature toggle or not. I think in general feature flagged code shouldn't move because it has a feature flag or not. (maybe this is different for permanent feature flags though? ie. thing that will never be generally available?)

I do agree that `routes/web.php` is too hard to scan though. Would it work to just split it into groups ie. routes/user, routes/form, routes/post etc? (that should happen on its on PR though)

File: app/Http/Controllers/API/Users/SettingsController.php [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+	}
+
+	public function index(Request $request)
+	{
+		$params = $this->getRouteParams($request);
+		
+		$this->usecase = $this->usecaseFactory
+			->get($this->getResource(), 'search');
+		
+		if (is_array($params) && array_key_exists('user_id', $params)) {
```
Comment (rjmackay) [inline app/Http/Controllers/API/Users/SettingsController.php]: This isn't necessary because `user_id` isn't optional right?

File: migrations/20180508154706_add_user_settings_table.php [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+use Phinx\Migration\AbstractMigration;
+
+class AddUserSettingsTable extends AbstractMigration
+{
+    /**
+    * Migrate Up.
+    */
+    public function up()
+    {
+        $this->table('user_settings')
```
Comment (rjmackay) [inline migrations/20180508154706_add_user_settings_table.php]: This needs a unique key constraint ie. you can't have duplicate config_key's for the same user right?

File: routes/features.php [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+		$router->group([
+			'prefix' => 'settings',
+			'middleware' => ['feature:user-settings']
+		], function () use ($router) {
+			$router->get('/', 'SettingsController@index');
+			$router->get('/{id:[0-9]+}', 'SettingsController@show');
+			$router->post('/', 'SettingsController@store');
+			$router->put('/{id:[0-9]+}', 'SettingsController@update');
+			$router->delete('/{id:[0-9]+}', 'SettingsController@destroy');
+		});
```
Comment (rjmackay) [inline routes/features.php]: I'm not sure about the structure of this. I think it matches the spec but may have issues.

1. I think ideally user settings shouldn't have IDs, rather user user_id+config_key as the primary key. The urls become `user/:user/settings/:config_key` instead.
2. I kind of think we should allow a create via PUT in this instance ie. `PUT user/1/settings/hxl_api_key` either creates or overwrites the existing API key. But doesn't really care if it existed already or not. I think that will match to how its used in the client

File: routes/web.php [hunk 14]
```diff
@@ -471,13 +467,23 @@
...
     $router->group([
+        'namespace' => 'Users',
         'prefix' => 'users',
         'middleware' => ['scope:users']
     ], function () use ($router) {
         // Public access
         $router->get('/', 'UsersController@index');
         $router->get('/{id:[0-9]+}', 'UsersController@show');
 
+        // Sub-user routes
```
Comment (rjmackay) [inline routes/web.php]: Why is this here, but repeated in features routes?

File: src/App/Init.php [hunk 15]
```diff
@@ -627,6 +634,12 @@
 $di->params[Ushahidi\App\Validator\User\Reset::class] = [
 	'repo'    => $di->lazyGet('repository.user')
 ];
+$di->params[Ushahidi\App\Validator\User\Setting\Update::class] = [
+	'user_repo'    => $di->lazyGet('repository.user')
+];
+$di->params[Ushahidi\App\Validator\Contact\Update::class] = [
```
Comment (rjmackay) [inline src/App/Init.php]: Why did this change in this PR?

File: src/App/Repository/Config/features.php [hunk 16]
```diff
@@ -85,4 +85,9 @@
 	'hxl' => [
 		'enabled' => false,
 	],
+
+	// Enable or disable User Settings feature
+	'user-settings' => [
+		'enabled' => true,
```
Comment (rjmackay) [inline src/App/Repository/Config/features.php]: This should be disabled by default. We need to add a way to override per environment but for now we should always be keeping things disabled unless we want them on in production.

File: src/App/Repository/User/SettingRepository.php [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+	{
+		$query = $this->search_query;
+
+		if ($search->user_id) {
+			$query->where('user_id', '=', $search->user_id);
+		}
+	}
+
+	// UserSettingRepository
+	public function updateCollection(array $entities)
```
Comment (rjmackay) [inline src/App/Repository/User/SettingRepository.php]: Is this used anywhere yet? I really like this idea ie. that you can overwrite all user settings at once, since that seems likely how it'll get used in the UI anyway

File: app/Http/Controllers/API/Users/SettingsController.php [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+			$this->usecase
+				->setFilters($request->query() + [
```
Comment (rjmackay) [inline app/Http/Controllers/API/Users/SettingsController.php]: I wonder if we could rewrote this slightly and avoid treating user_id as a filter... its really an identifier ie. part of the primary key, not an optional filter

File: src/App/Repository/User/SettingRepository.php [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+	// UserSettingRepository
+	public function getByUser($user_id)
+	{
+		$query = $this->selectQuery(compact($user_id));
+		$results = $query->execute($this->db);
+
+		return $this->getCollection($results->as_array());
+	}
+
+	public function deleteAllForForm($user_id)
```
Comment (rjmackay) [inline src/App/Repository/User/SettingRepository.php]: deleteAllForUser?

File: src/App/Validator/User/Setting/Update.php [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+				['digit'],
+				[[$this->user_repo, 'exists'], [':value']],
+			],
+			'config_key' => [
```
Comment (rjmackay) [inline src/App/Validator/User/Setting/Update.php]: Add a list of allowed keys? or maybe we don't even care

File: src/Core/Entity/UserSetting.php [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+    protected $config_key;
+    protected $config_value;
+    protected $created;
+    protected $updated;
+
+	// DataTransformer
+	protected function getDefinition()
+	{
+		return [
+			'id'       => 'int',
```
Comment (rjmackay) [inline src/Core/Entity/UserSetting.php]: I think ideally these don't have an ID, just a combined config_key+user_id as primary key

File: src/Core/Entity/UserSettingRepository.php [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+ * @copyright  2018 Ushahidi
+ * @license    https://www.gnu.org/licenses/agpl-3.0.html GNU Affero General Public License Version 3 (AGPL3)
+ */
+
+namespace Ushahidi\Core\Entity;
+
+use Ushahidi\Core\Entity\Repository\EntityGet;
+use Ushahidi\Core\Entity\Repository\EntityExists;
+
+interface UserSettingRepository extends
```
Comment (rjmackay) [inline src/Core/Entity/UserSettingRepository.php]: I haven't done this everywhere yet but this should probably extend Create and Update Repository interfaces I think

File: src/Core/Tool/Authorizer/UserSettingAuthorizer.php [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+		// These checks are run within the user context.
+		$user = $this->getUser();
+
+		// Only logged in users have access if the deployment is private
+		if (!$this->canAccessDeployment($user)) {
+			return false;
+		}
+
+		// Role with the Manage Users permission can manage all users
+		if ($this->acl->hasPermission($user, Permission::MANAGE_USERS)) {
```
Comment (rjmackay) [inline src/Core/Tool/Authorizer/UserSettingAuthorizer.php]: I think maybe this shouldn't be true. ie. a user should be the only person who can edit their own settings? Admins don't need to be able to edit a users settings

File: bootstrap/lumen.php (hunk 8 above)
Comment (willdoran) [reply bootstrap/lumen.php]: Yeah, splitting per endpoint would be ideal. We'll need a ticket for it,

Comment (rowasc) [reply bootstrap/lumen.php]: Agree on splitting per endpoints, that'd be nice. We could keep the CheckFeature and separate everything per endpoint , instead of having the routes/feature.php

Comment (rowasc) [reply bootstrap/lumen.php]: but another option is that everything has their own file , unless feature flagged, I guess? I kind of like the idea of being able to check which flags are still being used easily.

File: app/Http/Controllers/API/Users/SettingsController.php (hunk 11 above)
Comment (willdoran) [reply app/Http/Controllers/API/Users/SettingsController.php]: It was at one point but I think I removed that route. I'll check.

File: src/App/Init.php (hunk 15 above)
Comment (willdoran) [reply src/App/Init.php]: It was changed because of fiddling with white space in editors.

File: src/App/Repository/User/SettingRepository.php (hunk 17 above)
Comment (willdoran) [reply src/App/Repository/User/SettingRepository.php]: No it's only a placeholder so far.

File: src/App/Validator/User/Setting/Update.php (hunk 20 above)
Comment (willdoran) [reply src/App/Validator/User/Setting/Update.php]: Right now we don't have a list but I'll add it as a comment.

File: src/Core/Entity/UserSettingRepository.php (hunk 22 above)
Comment (willdoran) [reply src/Core/Entity/UserSettingRepository.php]: where are those?

File: src/Core/Tool/Authorizer/UserSettingAuthorizer.php (hunk 23 above)
Comment (willdoran) [reply src/Core/Tool/Authorizer/UserSettingAuthorizer.php]: Not sure about that. Ideally, you should be able to see all user settings but I'm not sure really what benefit it is. I'll remove it.

File: routes/web.php (hunk 14 above)
Comment (willdoran) [reply routes/web.php]: I checked I don't see it repeated but I was wondering why is it outside auth?

File: routes/web.php (hunk 1 above)
Comment (rjmackay) [reply routes/web.php]: Everything being added for HXL (or anything else) should get a feature flag. That way we can eventually deploy it to prod but disabled till its 100% working. I would argue for just using the HXL feature flag - on the assumption that this feature flag is only going to be used to disable the feature during develop and QA, not a long lived feature flag that will be tied to users plans.

File: src/Core/Tool/Authorizer/UserSettingAuthorizer.php [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+		// These checks are run within the user context.
+		$user = $this->getUser();
+
+		// Only logged in users have access if the deployment is private
+		if (!$this->canAccessDeployment($user)) {
+			return false;
+		}
+
+		// Admin user should be able to do anything - short of deleting self
+		if ($this->isUserAdmin($user)) {
```
Comment (rjmackay) [inline src/Core/Tool/Authorizer/UserSettingAuthorizer.php]: Sorry I wasn't clear before. I don't think this should be here either.

File: src/Core/Entity/UserSettingRepository.php (hunk 22 above)
Comment (rjmackay) [reply src/Core/Entity/UserSettingRepository.php]: `Ushahidi\Core\Usecase` I think

File: routes/web.php (hunk 14 above)
Comment (rjmackay) [reply routes/web.php]: Users are available to anonymous users to provide the users name for post detail - but we filter out email and other sensitive info.

File: migrations/20180508154706_add_user_settings_table.php [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+            ->addColumn('config_key', 'string', [
```
Comment (rjmackay) [inline migrations/20180508154706_add_user_settings_table.php]: Can you add an index since we'll be searching by config key?

Comment (willdoran) [reply migrations/20180508154706_add_user_settings_table.php]: Yep I have it's just not committed just yet.

**Comment by willdoran** [comment]
> I'm going to add groups as well.
What we're to do is basically have a bag of properties that belongs to a user, the likely operations we'll have are. Get me all settings by user or get me all settings related to a feature. I also don't really like config, I agree that the serialised json is awful and it solves a problem we don't really have which is trying to avoid millions of rows of settings data.

My major concern is that the keys are not going to be unique across different systems. We can make them unique ourselves on purpose but it would be much nicer to have the settings match the expected structure of either internal or external usecases. I'd rather be able to do 

`$config = User.getSettingFor('twitter');`

Than have to do 

`
$user_settings = User.getSettingFor('twitter');
$config = [
   'api_key' => $user_settings['twitter_api_key'],
....
   'some_setting_99 => $user_setting['twitter_some_setting_99']
`

The stupider the API can be about the data model the better. It doesn't really matter at this stage so I'll drop the group but it'll be a very annoying change to make later. Though I guess it could be looked at under data model. The important thing to recognise is that these aren't our settings they are at the moment 3rd party settings so we don't control the naming scheme.

File: src/Core/Entity/UserSettingRepository.php (hunk 22 above)
Comment (willdoran) [reply src/Core/Entity/UserSettingRepository.php]: I think we can do that in a separate ticket, can you add one for it and assign it to @jrtricafort so he can track it into a later cycle?

File: src/App/Init.php (hunk 15 above)
Comment (willdoran) [reply src/App/Init.php]: Our white space is wrong everywhere.
